### ISSUE TITLE
Can I use ranger22 and cosine annealing strategy together?

### ISSUE THREAD

**Issue by wudizuixiaosa** [lessw2020/Ranger22#1]
Hello, first of all, I'm very glad to see the emergence of ranger2022, which I'm most looking forward to. Thank you for your work.
I've been using your Ranger optimizer before, and its effect is very good, but I've been using mmdet to do some work on target detection recently. I don't know how to register Ranger's strategy into mmdet, so I want to ask whether ranger22 and cosine strategy can be used to achieve the effect。

**Comment by AdrienCourtois** [comment]
Hello, to use cosine annealing you only have to define the Scheduler as:
```scheduler = Scheduler21(optimizer, len(train_loader), epochs, warmdown_type="cosine")```
Don't forget to call ```scheduler.step()``` after each ```optimizer.step()```!

You can check the evolution of the learning rate with the following snippet:
```python
import pylab as plt
learning_rates = []

for epoch in range(1, epochs + 1):
  for x, y in train_loader:
    ## your code here

    optimizer.step()
    scheduler.step()

    for group in optimizer.param_groups:
      learning_rates.append(group['lr'])

plt.semilogy(learning_rates)
plt.show()
```
Note that additional tweaks might be necessary if you use different param groups, but that's not the default user's behavior.

**Comment by wudizuixiaosa** [comment]
Thank you very much for your reply, but in fact, this is the case
```python
lr_ config = dict(
 _ delete_= True,
 policy='CosineAnnealing',
 min_ lr_ ratio=0.01,
 warmup='linear',
 warmup_ iters=2000,
 warmup_ ratio=0.001)
 runner = dict(type='EpochBasedRunner', max_epochs=25)
optimizer = dict(
_ delete_= True, type='AdamW', lr=0.0001, weight_ decay=0.05,
paramwise_ cfg=dict(
custom_ keys={
'pos_ embed': dict(decay_mult=0.),
'norm': dict(decay_mult=0.),
'bias': dict(decay_mult=0.)
}))
```

This is a section in a config file in mmdet. I don't see any change methods for setting training strategies. I can only register Ranger with the optimizer like this
```python
lr_config = dict(
    _delete_=True,
    policy='CosineAnnealing',
    min_lr_ratio=0.01,
    warmup='linear',
    warmup_iters=2000,
    warmup_ratio=0.001)
runner = dict(type='EpochBasedRunner', max_epochs=50)
optimizer = dict(type='ranger22')
```

**Comment by AdrienCourtois** [comment]
I am not familiar with mmdet. Perhaps you should ask in their repo how to use a custom scheduler.

**Comment by wudizuixiaosa** [comment]
OK, thank you for your generous answer

**Comment by wudizuixiaosa** [comment]
> custom scheduler.

By the way，Avoid my misunderstanding of ranger22，I want to ask，Is this the right way to use it
from ranger22 import Ranger22
 optimizer = Ranger22(model.parameters())
and i just adjust the lr in the ranger22.py。
because I haven't figured out the most correct way to use Ranger, haha。

**Comment by AdrienCourtois** [comment]
If you want to use the optimizer and the scheduler, the correct way to do so would be:
```python
from ranger22 import Ranger22
from scheduler22 import Scheduler22

optimizer = Ranger22(model.parameters(), lr=5e-3)
scheduler = Scheduler21(optimizer, epochs, len(train_loader))

for epoch in range(1, epochs + 1):
  for x, y in train_loader:
    # Forward pass

    optimizer.step()
    scheduler.step()
```

**Comment by wudizuixiaosa** [comment]
> If you want to use the optimizer and the scheduler, the correct way to do so would be:
> 
> ```python
> from ranger22 import Ranger22
> from scheduler22 import Scheduler22
> 
> optimizer = Ranger22(model.parameters(), lr=5e-3)
> scheduler = Scheduler21(optimizer, epochs, len(train_loader))
> 
> for epoch in range(1, epochs + 1):
>   for x, y in train_loader:
>     # Forward pass
> 
>     optimizer.step()
>     scheduler.step()
> ```
I'm very sorry. I thought it could be solved, but I can only ask you for help again.
https://github.com/facebookresearch/mae/blob/main/main_finetune.py
I want to try to add Ranger to some existing finetune projects, such as the MAE finetune above.
I lived according to your premise
    optimizer = Ranger22(model.parameters(), lr=5e-3)
    scheduler = Scheduler22(optimizer, args.epochs, len(data_loader_train))

    for args.epoch in range(1, args.epochs + 1):
        for x, y in data_loader_train:
            # Forward pass

            optimizer.step()
            scheduler.step()
but it doesnot work，Can you help me see how to modify it, at your convenience.
by the way，If I only import 
 optimizer = Ranger22(model.parameters(), lr=5e-3)
it can work，but The result is not as good as the original adamw。

**Comment by AdrienCourtois** [comment]
You will have to be more specific: can you please report the error you have?

**Comment by wudizuixiaosa** [comment]
My modified version is as follows，https://github.com/wudizuixiaosa/finetune/blob/main/.gitattributes
when i run it，it will be stuck after outputting the network structure and will not continue

**Comment by wudizuixiaosa** [comment]
In fact, I feel that I failed to call scheduler successfully, which made the process unable to proceed, but I don't know how to change it, so I hope you can see the code I modified if it's convenient. I just imported and modified the parameter name in your way, but it can't work
